curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.md

```diff
@@ -85,33 +85,128 @@ tests:
+        assert(data.error === 'invalid number' || data === 'invalid number');
+      } catch(xhr) {
+        throw new Error(xhr.responseText || xhr.message);
+      }
+    }"
+  - text: If both the unit and number are invalid, returned will be <code>'invalid number and unit'</code>.
+    testString: "async getUserInput => { 
+      try {
+        const data = await $.get(getUserInput('url') + '/api/convert?input=1//2min');
+        assert(data.error === 'invalid number and unit' || data === 'invalid number and unit');
+      } catch(xhr) {
+        throw new Error(xhr.responseText || xhr.message);
+      }
+    }"
   - text: I can use fractions, decimals or both in my parameter(ie. 5, 1/2, 2.5/6), but if nothing is provided it will default to 1.
-    testString: ''
-  - text: My return will consist of the initNum, initUnit, returnNum, returnUnit, and string spelling out units in format '{initNum} {initial_Units} converts to {returnNum} {return_Units}' with the result rounded to 5 decimals in the string.
-    testString: ''
+    testString: "async getUserInput => {
+      try {
+        const data1 = await $.get(getUserInput('url') + '/api/convert?input=mi');
+        assert.approximately(data1.initNum, 1, 0.001);
+        assert.approximately(data1.returnNum, 1.60934, 0.001);
+        assert.equal(data1.returnUnit, 'km');
+        const data2 = await $.get(getUserInput('url') + '/api/convert?input=1/5mi');
+        assert.approximately(data2.initNum, 1/5, 0.1);
+        assert.approximately(data2.returnNum, 0.32187, 0.001);
+        assert.equal(data2.returnUnit, 'km');
+        const data3 = await $.get(getUserInput('url') + '/api/convert?input=1.5/7km');
+        assert.approximately(data3.initNum, 1.5/7, 0.001);
+        assert.approximately(data3.returnNum, 0.13315, 0.001);
+        assert.equal(data3.returnUnit, 'mi');
+        const data4 = await $.get(getUserInput('url') + '/api/convert?input=3/2.7km');
+        assert.approximately(data4.initNum, 3/2.7, 0.001);
+        assert.approximately(data4.returnNum, 0.69041, 0.001);
+        assert.equal(data4.returnUnit, 'mi');
+      } catch(err) {
+        throw new Error(err.responseText || err.message);
+      }
+    }
+    "
+  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code> spelling out units in the format <code>'{initNum} {initial_Units} converts to {returnNum} {return_Units}'</code> with the result rounded to 5 decimals.
+    testString: "async getUserInput => { 
+      try {
+        const data = await $.get(getUserInput('url') + '/api/convert?input=2mi');
+        assert.equal(data.initNum, 2);
+        assert.equal(data.initUnit, 'mi');
+        assert.approximately(data.returnNum, 3.21868, 0.001);
+        assert.equal(data.returnUnit, 'km', 'returnUnit did not match');
+        assert.equal(data.string, '2 miles converts to 3.21868 kilometers')
```

Comment:
To clarify the issue I am seeing here:
- `initial_units` and `return_units` are not referenced before this
- I missed the HTML example output (when I did this project), and followed this pattern:
```
{initNum} {initUnit} converts to {returnNum} {returnUnit}
```

I see two options:
1) Change the tests and HTML example output to not need the _Full-English_ unit name
2) Make the challenge more difficult with this user story as is, but be clear (somewhere else) that there is a requirement to return the _Full-English_ unit names...at some point

Reply:
```yml
  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code>. <code>string</code> will take the format <code>'{initNum} {initUnit} converts to {returnNum} {returnUnit}'</code> with the converted result rounded to 5 decimals, and <code>initUnit</code> and <code>returnUnit</code> values should be spelled out (e.g. `km` becomes `kilometer`).
```
I do lean more toward keeping the requirement - how is this?

Reply:
@nhcarrigan Almost. I would not have `initUnit` and `returnUnit` be the same (namespace) as the others. Maybe even something more verbose: `initUnitString & returnUnitString`

Reply:
That's a good approach, I think.